### DISCUSSION TITLE
Azure DB using Managed Identity

### DISCUSSION THREAD

**Discussion by gaclique** [nhibernate/nhibernate-core#2694] [Q&A]
Has anyone had any luck configuring nhibernate to connect to Azure SQL db using Managed Identity.  This removes the username / password from the connection string altogether and uses an access token retrieved from the Microsoft.Azure.Services.AppAuthentication library.  Any insights into this would be greatly appreciated.

**Comment by gliljas** [comment]
You should be able to accomplish this by creating your own ConnectionProvider.

**Reply by gaclique** [reply]
Yes that is what I decided to do.  I just thought that the current sql driver would support it but it doesn’t appear so

**Reply by gliljas** [reply]
Supporting it in the standard driver would require unwanted dependencies. As a separate project it would be worthwhile.

**Reply by mthadisena** [reply]
> Yes that is what I decided to do. I just thought that the current sql driver would support it but it doesn’t appear so

I'm getting this error: "Login failed for user ''. I'm using 4.0.0. Which version of nHibernate are you using?

**Comment by hazzik** [comment]
What database are you talking about? If it is SQL Server, then `Microsoft.Data.SqlClient` does support Managed Identities out of the box. MySQL and PostgreSQL would require a custom code.

**Reply by gaclique** [reply]
Yes, both the Microsoft.Data.SqlClient and System.Data.SqlClient does support it out of the box.  I was referring to the nHibernate.Driver.SqlClientDriver having support.  But as @gliljas mentioned above, that would require unwanted dependencies.

**Reply by hazzik** [reply]
You would not need changes to `NHibernate.Driver.SqlClientDriver` to support MI with SQL Server. What you'd need is to change the connection string.

**Reply by gliljas** [reply]
I agree. My comment was referring to the use of AppAuthentication and setting the token. That is no longer necessary (which was news to me, but it was some time ago I used Azure SQL), so you should be good to go with an amended connectionstring  and a modern version of SqlClient.

**Comment by mthadisena** [comment]
> Has anyone had any luck configuring nhibernate to connect to Azure SQL db using Managed Identity. This removes the username / password from the connection string altogether and uses an access token retrieved from the Microsoft.Azure.Services.AppAuthentication library. Any insights into this would be greatly appreciated.

Which version of nhibernate are you using? I'm using 4.0.0.  and using System.Data.SqlClient 4.0.0. I'm getting error "Login failed for user ''". Here's the code:
--------------------------
var sqlConnection = new SqlConnection(connectionString);
var credential = new DefaultAzureCredential();
var token = credential.GetToken(new TokenRequestContext(new[] { "https://database.windows.net/.default" }));
sqlConnection.AccessToken = token.Token;
var session = SessionManager.OpenSession(sqlConnection);
